src/target/riscv/riscv-013.c

```diff
@@ -3396,310 +3396,524 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
-				buf_set_u64(buffer + (next_index - 2) * size, 0, 8 * size, value64);
-				log_memory_access64(address + (next_index - 2) * size, value64, size, true);
+	uint32_t elements_to_fetch_from_batch;
 
-				/* Restore the command, and execute it.
-				 * Now DM_DATA0 contains the next value just as it would if no
-				 * error had occurred. */
-				dmi_write_exec(target, DM_COMMAND, command, true);
-				next_index++;
+	info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
+	switch (info->cmderr) {
+	case CMDERR_NONE:
+		LOG_TARGET_DEBUG(target, "successful (partial?) memory read [%"
+				PRIu32 ", %" PRIu32 ")", start_index, start_index + elements_to_read);
+		elements_to_fetch_from_batch = elements_to_read;
+		break;
+	case CMDERR_BUSY:
+		LOG_TARGET_DEBUG(target, "memory read resulted in busy response");
+		if (read_memory_progbuf_inner_on_ac_busy(target, start_index,
+					elements_to_read, &elements_to_fetch_from_batch, access)
+				!= ERROR_OK)
+			return ERROR_FAIL;
+		break;
+	default:
+		LOG_TARGET_DEBUG(target, "error when reading memory, cmderr=0x%" PRIx32,
+				info->cmderr);
+		riscv013_clear_abstract_error(target);
+		return ERROR_FAIL;
+	}
 
-				dmi_write(target, DM_ABSTRACTAUTO,
-						1 << DM_ABSTRACTAUTO_AUTOEXECDATA_OFFSET);
+	if (read_memory_progbuf_inner_extract_batch_data(target, batch, start_index,
+				elements_to_fetch_from_batch, elements_read, access) != ERROR_OK)
+		return ERROR_FAIL;
 
-				ignore_last = 1;
+	return ERROR_OK;
+}
 
-				break;
-			default:
-				LOG_DEBUG("error when reading memory, abstractcs=0x%08lx", (long)abstractcs);
-				riscv013_clear_abstract_error(target);
-				riscv_batch_free(batch);
-				result = ERROR_FAIL;
-				goto error;
-		}
+static uint32_t read_memory_progbuf_inner_fill_batch(struct riscv_batch *batch,
+		uint32_t count, uint32_t size)
+{
+	assert(size <= 8);
+	const uint32_t two_regs_used[] = {DM_DATA1, DM_DATA0};
+	const uint32_t one_reg_used[] = {DM_DATA0};
+	const uint32_t reads_per_element = size > 4 ? 2 : 1;
+	const uint32_t * const used_regs = size > 4 ? two_regs_used : one_reg_used;
+	const uint32_t batch_capacity = riscv_batch_available_scans(batch) / reads_per_element;
+	const uint32_t end =  MIN(batch_capacity, count);
+
+	for (uint32_t j = 0; j < end; ++j)
+		for (uint32_t i = 0; i < reads_per_element; ++i)
+			riscv_batch_add_dmi_read(batch, used_regs[i]);
+	return end;
+}
 
-		/* Now read whatever we got out of the batch. */
-		dmi_status_t status = DMI_STATUS_SUCCESS;
-		unsigned read_count = 0;
-		assert(index >= 2);
-		for (unsigned j = index - 2; j < index + reads; j++) {
-			assert(j < count);
-			LOG_DEBUG("index=%d, reads=%d, next_index=%d, ignore_last=%d, j=%d",
-				index, reads, next_index, ignore_last, j);
-			if (j + 3 + ignore_last > next_index)
-				break;
+static int read_memory_progbuf_inner_try_to_read(struct target *target,
+		struct memory_access_info access, uint32_t *elements_read,
+		uint32_t index, uint32_t loop_count)
```

Comment:
Maybe `loop_count` could be renamed to something more descriptive. E.g. `max_elements_to_read` ?

```suggestion
		uint32_t index, uint32_t max_elements_to_read)
```

Reply:
I have tried to come up with a better name.

IMHO, renaming loop_count to max_elements_to_read would make it seem like this is the maximum value of elements_to_read, but it is not the case -- elements_to_read are less or equal to loop_count - index.

The other approach is something like max_index, but in such case it is easy to assume that this maximum is reached by index, meaning on some iteration index is equal to loop_count, which is not the case -- it is always less and the loop ends when index == loop_count - 1.

If you have any other suggestions, I'd be glad to hear them.

Reply:
Thanks for explanation. In that case, let's leave it as is :)